### PR TITLE
#260 Capability to calculate percentages in Classification

### PR DESCRIPTION
**Pull request by taba90**
Issue https://github.com/geosolutions-it/jai-ext/issues/260

Percentages to the [EqualAreaBreaksOpImage ](https://github.com/taba90/jai-ext/blob/4482ff2b4a1c9b3dc1b7bddf31079f4b088b88a3/jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/EqualIntervalBreaksOpImage.java)class
have not been added due to the fact that geoserver compute equal intervals inside sld module

### CODE REVIEW

File: jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/NaturalBreaksHistogramOpImage.java [hunk 1]
```diff
@@ -166,5 +169,25 @@ protected void postCalculate(Classification c, int band) {
         }
         breaks[0] = buckets.get(0).getMin();
         hc.setBreaks(band, breaks);
+        if(percentages.booleanValue())
```
Comment (aaime) [inline jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/NaturalBreaksHistogramOpImage.java]: jai-ext does not have auto-formatting, but could you still try to be consistent with gt/gs?

File: jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/NaturalBreaksHistogramOpImage.java [hunk 2]
```diff
@@ -166,5 +169,25 @@ protected void postCalculate(Classification c, int band) {
...
+    private double[] getPercentages(List<Bucket> buckets, int m, Double[] breaks) {
+        double [] percentages = new double [numClasses];
+        double [] values = buckets.stream().mapToDouble(b->b.getMin()).toArray();
+        for (int i = 0; i< numClasses; i++){
+            double break1 = breaks[i];
+            double break2 = breaks[i+1];
+            boolean last = i==(numClasses-1);
+            double classMembers;
+            if (last){
+                classMembers= DoubleStream.of(values).filter(v->v>=break1 && v<=break2).count();
```
Comment (aaime) [inline jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/NaturalBreaksHistogramOpImage.java]: Usage of streams normally provides lower performance compared to loops, please avoid them in high volume situations and use a classic array for loop. JAI development requires a C programmer mindset.

File: jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/NaturalBreaksOpImage.java [hunk 3]
```diff
@@ -159,5 +161,23 @@ protected void postCalculate(Classification c, int band) {
...
+    }
+
+    private double[] getPercentages(List<Double> data, int m, Double[] breaks) {
+        double [] percentages = new double [numClasses];
+        for (int i = 0; i< numClasses; i++){
+            double break1 = breaks[i];
+            double break2 = breaks[i+1];
+            double classMembers;
+            if (i+1==numClasses){
+                classMembers= data.stream().filter(v->v>=break1 && v<=break2).count();
```
Comment (aaime) [inline jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/NaturalBreaksOpImage.java]: Same observation as above regarding streams

File: jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/QuantileBreaksOpImage.java [hunk 4]
```diff
@@ -116,6 +117,19 @@ protected void postCalculate(Classification c, int band) {
...
+        qc.setBreaks(band, set.toArray(new Double[actualClassNum]));
+        if (this.percentages.booleanValue())
+            qc.setPercentages(getPercentages(dSize, actualClassNum));
+    }
+
+    private double[] getPercentages (double nvalues,int classNum){
+        double size = nvalues / (double) classNum;
+        double [] percentages = new double[classNum];
+        for (int i=0; i<classNum; i++){
+            percentages[i]= (size/nvalues)*100;
```
Comment (aaime) [inline jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/QuantileBreaksOpImage.java]: Quantile tries to build equal classes, data might disagree and make it impossible, please add a test with a raster having only 3 values, 5 classes, and one of the values taking half of the pixels, while the other 2 are getting 25% each. Check that the result is correct and percentages are too.

**Comment by aaime** [comment]
@taba90 this is a stand alone library, not a tool specifically designed to work with the SLD service. Please either make the code work consistently, or make it throw an exception explaining why percentages are not supported in EqualIntervalBreaksOpImage (EqualAreaBreaksOpImage I believe does not exist, your link pointed to the equal interval one).

File: jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/NaturalBreaksHistogramOpImage.java (hunk 1 above)
Comment (taba90) [reply jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/NaturalBreaksHistogramOpImage.java]: fixed

File: jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/NaturalBreaksHistogramOpImage.java (hunk 2 above)
Comment (taba90) [reply jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/NaturalBreaksHistogramOpImage.java]: fixed

File: jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/NaturalBreaksOpImage.java (hunk 3 above)
Comment (taba90) [reply jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/NaturalBreaksOpImage.java]: fixed

File: jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/QuantileBreaksOpImage.java (hunk 4 above)
Comment (taba90) [reply jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/QuantileBreaksOpImage.java]: fixed

**Comment by taba90** [comment]
Thanks for the review. I've added also percentages computation for EqualInterval. I made EqualIntervalBreaksOpImage relying on the NaturalClassification class if percentages are required in order to have values available for percentages computing.

File: jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/NaturalBreaksHistogramOpImage.java [hunk 5]
```diff
@@ -166,5 +167,21 @@ protected void postCalculate(Classification c, int band) {
...
         breaks[0] = buckets.get(0).getMin();
         hc.setBreaks(band, breaks);
+        if (percentages.booleanValue()) {
+            ClassPercentagesManager percentagesManager = new ClassPercentagesManager();
+        }
+            hc.setPercentages(
+                    getPercentages(buckets, m, k>breaks.length?breaks.length:k, breaks));
+    }
+
+    private double[] getPercentages(List<Bucket> buckets, int m, int k, Double[] breaks) {
```
Comment (aaime) [inline jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/NaturalBreaksHistogramOpImage.java]: I don't understand how this is working... each bucket represents a range and the count of values in that range, no?
If you don't use the counts from the buckets, how are percentages computed?

File: jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/NaturalBreaksHistogramOpImage.java [hunk 6]
```diff
@@ -166,5 +167,21 @@ protected void postCalculate(Classification c, int band) {
         }
         breaks[0] = buckets.get(0).getMin();
         hc.setBreaks(band, breaks);
+        if (percentages.booleanValue()) {
```
Comment (aaime) [inline jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/NaturalBreaksHistogramOpImage.java]: This is istantiates a dangling percentagesManager that does not get used?

File: jt-classbreaks/src/test/java/it/geosolutions/jaiext/classbreaks/ClassBreaksOpImageTest.java [hunk 7]
```diff
@@ -204,6 +199,154 @@ public void testNaturalBreaksHistogram() throws Exception {
...
+        Classification classification =
+                (Classification) op.getProperty(ClassBreaksDescriptor.CLASSIFICATION_PROPERTY);
+        assertNotNull(classification);
+        double[] percentages = classification.getPercentages();
+        assertEquals(percentages.length, 4);
+        assertTrue(percentages[0]== 18.75);
+        assertTrue(percentages[1]== 31.25);
+        assertTrue(percentages[2]== 25.0);
+        assertTrue(percentages[3]== 25.0);
+        // Test required classes > breaks
```
Comment (aaime) [inline jt-classbreaks/src/test/java/it/geosolutions/jaiext/classbreaks/ClassBreaksOpImageTest.java]: One test method, one test, please split and give the method a name that talks about what the test is doing. Do the same for all other tests.

File: jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/NaturalBreaksHistogramOpImage.java (hunk 5 above)
Comment (taba90) [reply jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/NaturalBreaksHistogramOpImage.java]: fixed

File: jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/NaturalBreaksHistogramOpImage.java (hunk 6 above)
Comment (taba90) [reply jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/NaturalBreaksHistogramOpImage.java]: fixed

File: jt-classbreaks/src/test/java/it/geosolutions/jaiext/classbreaks/ClassBreaksOpImageTest.java (hunk 7 above)
Comment (taba90) [reply jt-classbreaks/src/test/java/it/geosolutions/jaiext/classbreaks/ClassBreaksOpImageTest.java]: fixed

File: jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/ClassPercentagesManager.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        for (int i = 0; i < buckets.size(); i++) {
+            totalSize += buckets.get(i).getCount();
+        }
+        int bucketSize = buckets.size();
+        for (int i = 0; i < numClasses; i++) {
+            double current = breaks[i];
+            double next = breaks[i + 1];
+            boolean last = numClasses == i + 1;
+            double classMembers;
+            if (last) {
```
Comment (aaime) [inline jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/ClassPercentagesManager.java]: Given the tests checking alternatively the max, middle or min value, it does not seem impossible to find buckets that match both criteria, e.g, whose min is below the current break, but whose average sits in the next class.
Perhaps it is wrong to try and assign a bucket fully into a class, a different approach would be to assume the bucket is uniformly distributed, and assign counts to the output ranges based on the percentage of overlap instead?

File: jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/QuantileBreaksHistogramOpImage.java [hunk 9]
```diff
@@ -102,8 +99,24 @@ protected void postCalculate(Classification c, int band) {
...
+        }
+    }
+
+    private double[] getPercentages(List<Double> tBreaks, List<Bucket> buckets, int nvalues, int numClasses) {
+        Map<Double, Integer> values = new HashMap<>();
+        for (Bucket b : buckets) {
+            values.put(b.getMin(), b.getCount());
+        }
+        ClassPercentagesManager percentagesManager = new ClassPercentagesManager();
+        return percentagesManager.getPercentages(values, tBreaks, nvalues, numClasses);
```
Comment (aaime) [inline jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/QuantileBreaksHistogramOpImage.java]: Both histogram based natural breaks and quantiles percentage calculation require counting portions of buckets in a given set of ranges, that do not necessarily match the buckets limits. So why are the two using two different methods to compute the results?

Comment (taba90) [reply jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/QuantileBreaksHistogramOpImage.java]: fixed

File: jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/ClassPercentagesManager.java (hunk 8 above)
Comment (taba90) [reply jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/ClassPercentagesManager.java]: fixed

File: jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/NaturalBreaksHistogramOpImage.java [hunk 10]
```diff
@@ -32,12 +32,11 @@
...
+import it.geosolutions.jaiext.classbreaks.HistogramClassification.Bucket;
 
 import javax.media.jai.ROI;
-
-import it.geosolutions.jaiext.classbreaks.HistogramClassification.Bucket;
+import java.awt.image.RenderedImage;
+import java.util.List;
 
 /** Classification op for the natural breaks method. */
 public class NaturalBreaksHistogramOpImage extends ClassBreaksOpImage {
```
[the remark this answers is not in the document]
Comment (taba90) [reply jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/NaturalBreaksHistogramOpImage.java]: fixed

File: jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/ClassPercentagesManager.java [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+            double current = breaks.get(i);
+            double next = breaks.get(i + 1);
+            boolean last = i + 1 == numClasses;
+            Double overlapping;
+            if (last) {
+                overlapping = getOverlappingPercentages(current, next, b.getMin(), b.getMax());
+            } else {
+                overlapping = getOverlappingPercentages(current, Math.nextDown(next), b.getMin(), b.getMax());
+            }
+            if (overlapping.doubleValue() == 1.0 || overlapping.equals(NaN)) {
```
[the remark this answers is not in the document]
Comment (taba90) [reply jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/ClassPercentagesManager.java]: fixed

File: jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/ClassPercentagesManager.java [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+
+
+    private Double getOverlappingPercentages(double cMin, double cMax,
+                                             double bMin, double bMax) {
+        if (cMin > bMax || cMax < bMin) {
+            return 0.0;
+        }
+        double regionMin = Math.max(cMin, bMin);
+        double regionMax = Math.min(cMax, bMax);
+        return (regionMax - regionMin) / (bMax - bMin);
```
[the remark this answers is not in the document]
Comment (taba90) [reply jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/ClassPercentagesManager.java]: fixed

File: jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/ClassPercentagesManager.java [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+    private Double getOverlappingPercentages(double cMin, double cMax,
+                                             double bMin, double bMax) {
+        if (cMin > bMax || cMax < bMin) {
+            return 0.0;
+        }
+        double regionMin = Math.max(cMin, bMin);
+        double regionMax = Math.min(cMax, bMax);
+        double dividend = regionMax - regionMin;
+        double divider = bMax - bMin;
+        if (dividend == divider)
```
Comment (aaime) [inline jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/ClassPercentagesManager.java]: But why not check explicitly for zero? This is opaque, it seems like you're trying to optimize away a division.

Comment (taba90) [reply jt-classbreaks/src/main/java/it/geosolutions/jaiext/classbreaks/ClassPercentagesManager.java]: fixed

**Pull request merged**